modules/kafka/tests/test_kafka_topic.c

```diff
@@ -0,0 +1,202 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include <syslog-ng.h>
+#include <apphook.h>
+#include "logthrdest/logthrdestdrv.h"
+#include "kafka-dest-driver.h"
+#include <librdkafka/rdkafka.h>
+#include "kafka-dest-worker.h"
+#include "kafka-dest-driver.h"
+#include "kafka-dest-worker-internal.h"
+
+#include <criterion/criterion.h>
+#include <criterion/parameterized.h>
+
+struct topic_validity_test_params
+{
+  gchar *topic_name;
+  gboolean isvalid;
+};
+
+static void
+setup(void)
+{
+  app_startup();
+}
+
+static void
+teardown(void)
+{
+  app_shutdown();
+}
+
+TestSuite(kafka_topic, .init = setup, .fini = teardown);
+
+ParameterizedTestParameters(kafka_topic, topic_validity_tests)
+{
+  static struct topic_validity_test_params params[] =
+  {
+    {"test 1", FALSE},
+    {"", FALSE},
+    {"..", FALSE},
+    {".", FALSE},
+    {" invalidname", FALSE},
+    {"a*123", FALSE},
+    {"topic@123", FALSE},
+    {"aba#2", FALSE},
+    {"dontuse&", FALSE},
+    {"illegal%", FALSE},
+    {"name!", FALSE},
+    {"topics^", FALSE},
+    {"round()", FALSE},
+    {"sum+", FALSE},
+    {"topic=name", FALSE},
+    {"topic``~invalid", FALSE},
+    {STRING_250_LEN, FALSE},
+    {"validname123", TRUE},
+    {"valid.name123", TRUE},
+    {"valid-name123", TRUE},
+    {"valid_name123", TRUE},
+    {"valid.name-123", TRUE},
+    {"valid_name-123", TRUE},
+    {"This.is.a.valid.name", TRUE},
+    {"This.is-a_valid-name", TRUE},
+    {STRING_249_LEN, TRUE},
+  };
+
+  return cr_make_param_array(struct topic_validity_test_params, params, sizeof(params)/sizeof(params[0]));
+}
+
+ParameterizedTest(struct topic_validity_test_params *param, kafka_topic, topic_validity_tests)
+{
+  KafkaDestDriver *driver = (KafkaDestDriver *) kafka_dd_new(configuration);
+  cr_assert_eq(kafka_dd_validate_topic_name(driver, param->topic_name), param->isvalid);
+  log_pipe_unref((LogPipe *)driver);
+}
+
+Test(kafka_topic, test_resolve_template_topic_name)
+{
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
+
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbacktopicname");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
+
+  LogMessage *msg = log_msg_new_empty();
+  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "valid_template_topic");
```

Comment:
maybe it's just me, but this would read much better if the function would use a KafkaDestWorker specific prefix (e.g. kafka_dest_worker_resolve_template_topic_name(), that would clearly indicate it is the function being tested. The names starting with an underscore is usually a private function not intended to be called from the outside.

The unit tests for the KafkaDestDriver are a bit clumsy, as we lack the internal layers so we could test these functions. However, the internal layering would be too much of boilerplate for not much gain.

With all this said, I think if you removed these functions to better indicate that they are part of the class being tested, this would be easier to understand.